lib/fog/compute/ecloud.rb

```diff
@@ -0,0 +1,822 @@
+            options.merge!({:body => params[:body]})
+          end
+          response = @connections[host_url].request(options)
+          # Parse the response body into a hash
+          unless response.body.empty?
+            if params[:parse]
+              document = Fog::ToHashDocument.new
+              parser = Nokogiri::XML::SAX::PushParser.new(document)
+              parser << response.body
+              parser.finish
+
+              response.body = document.body
+            end
+          end
+
+          response
+        end
+
+        private
+
+        # if Authorization and x-tmrk-authorization are used, the x-tmrk-authorization takes precendence.
+        def set_extra_headers_for(params)
+          params[:headers] = {
+            'x-tmrk-version' => @version,
+            'Date'           => Time.now.utc.strftime("%a, %d %b %Y %H:%M:%S GMT"),
+          }.merge(params[:headers] || {})
+          if params[:method]=="POST" || params[:method]=="PUT"
+            params[:headers].merge!({"Content-Type" => 'application/xml'}) unless params[:headers]['Content-Type']
+            params[:headers].merge!({"Accept" => 'application/xml'})
+          end
+          unless params[:body].nil? || params[:body].empty?
+            params[:headers].merge!({"x-tmrk-contenthash" => "Sha256 #{Base64.encode64(Digest::SHA2.digest(params[:body].to_s)).chomp}"})
+          end
+          if @authentication_method == :basic_auth
+            params[:headers].merge!({'Authorization' => "Basic #{Base64.encode64(@username+":"+@password).delete("\r\n")}"})
+          elsif @authentication_method == :cloud_api_auth
+            signature = cloud_api_signature(params)
+            params[:headers].merge!({
+              "x-tmrk-authorization" => %{CloudApi AccessKey="#{@access_key}" SignatureType="HmacSha256" Signature="#{signature}"},
+              "Authorization" => %{CloudApi AccessKey="#{@access_key}" SignatureType="HmacSha256" Signature="#{signature}"}
+            })
+          end
+          params[:headers]
+        end
+
+        def cloud_api_signature(params)
+          verb = params[:method].upcase
+          headers = params[:headers]
+          path = params[:uri].path
+          canonicalized_headers = canonicalize_headers(headers)
+          canonicalized_resource = canonicalize_resource(path)
+          string = [
+            verb,
+            headers['Content-Length'].to_s,
+            headers['Content-Type'].to_s,
+            headers['Date'].to_s,
+            canonicalized_headers,
+            canonicalized_resource + "\n"
+          ].join("\n")
+          Base64.encode64(@hmac.sign(string)).chomp
+        end
+
+        # section 5.6.3.2 in the ~1000 page pdf spec
+        def canonicalize_headers(headers)
+          tmp = headers.reduce({}) {|ret, h| ret[h.first.downcase] = h.last if h.first.match(/^x-tmrk/i) ; ret }
+          tmp.reject! {|k,v| k == "x-tmrk-authorization" }
```

Comment:
Unused block argument - `v`. If it's necessary, use `_` or `_v` as an argument name to indicate that it won't be used.<br>Space missing after comma.<br>Space between { and | missing.